Bc prop overhaul

NoSloppy opened the pull request.

NoSloppy reviewed (commented):
edit guards for scroll_presets_

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -2,82 +2,122 @@
 saber_BC_buttons.h
   http://fredrik.hubbe.net/lightsaber/proffieos.html
   Copyright (c) 2016-2019 Fredrik Hubinette
-  Copyright (c) 2023 Brian Conner with contributions by:
```
2023-2024

NoSloppy replied:
done

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -2,82 +2,122 @@
   http://www.gnu.org/licenses/
 
-Includes 1 and 2 button controls.
+Includes 1, or 2 button controls. (3rd button for power control of Dual Blades supported as well).
```
no comma

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -364,14 +1095,31 @@ Turn OFF without postoff - Turn OFF while pointing up.
+#define BC_SECOND_BLADE 2
+#endif
+
+#if BC_MAIN_BLADE == BC_SECOND_BLADE
```
Only check this if BC_DUAL_BLADES is defined

NoSloppy replied:
ok

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1134,317 @@ EFFECT(volmax);     // for maximum volume reached
+  int current_volume_ = 0;
+  int percentage_ = 0;
+
+int steps_per_revolution() override {
```
indentation (should be same as the line above)

NoSloppy replied:
thanks.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1134,317 @@ EFFECT(volmax);     // for maximum volume reached
-        trigger_time_ = millis();
-        duration_ = duration;
+  void mode_deactivate() override {
+    if (cancelling_) {
```
the "if cancelling" should be done in exit()

NoSloppy replied:
ah yes, no need for cancelling_ bool at all. done.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1134,317 @@ EFFECT(volmax);     // for maximum volume reached
-        return false;  // Timer is still running
+  void say() override {
+    if (last_volume_ < current_volume_) {
+      mode::getSL<SPEC>()->SayVolumeUp();
```
Why not do SayVolumeUp() in next() and SayVolumeDown() in Prev() ?

NoSloppy replied:
I wanted to, but they only speak if they are in say(). I do not understand why but nonetheless that is the case.

profezzorn replied:
Hmm, I guess that's because there is a fadeout() in update(), which is called right after next/prev.
I guess even the SteppedMode assumes that say() will announce the current state rather than speak to what action was taken.

It feels like the right thing to do here would be to split SteppedMode into two classes, something like SteppedModeBase and SteppedMode. SteppedModeBase would just provide next() prev(), but not make any assumptions about the sound, while SteppedMode would add update(), say() and all the other stuff that SteppedMode does....

Although, the current implementation works too I suppose...

NoSloppy replied:
That's fine since I'm overriding update() anyway so SayVolumeUp and Down can be used instead of ccchange, and there's no speaking at each step anyway, so by removing fadeout in this case allows speaking in next() and prev() tp work fine.

NoSloppy replied:
done

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1134,317 @@ EFFECT(volmax);     // for maximum volume reached
+
+#ifndef BC_ENABLE_OS_MENU
+template<class SPEC>
+struct BCVolumeMode : public SPEC::MenuBase {
```
If you're not going to use pos_ for anything, then you should probably just inherit from SteppedMode.
If you do want to use pos_, then you should set pos_ to something in mode_activate and and then calculate the volume from pos_ when pos_ changes.

NoSloppy replied:
no pos_ needed . inherit SteppedMode it is.

NoSloppy replied:
done

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1134,317 @@ EFFECT(volmax);     // for maximum volume reached
+#ifdef BC_DUAL_BLADES
+template<class SPEC>
+struct BCSelectBladeMode : public SPEC::MenuBase {
+  int current_blade_ = 1; // Start with the first blade
```
why not use pos_ +1 instead?

NoSloppy replied:
I don't know? I thought current_blade_ was very clear as to what it is.
Also, pos_ is not in scope.

profezzorn replied:
Because C++ is weird, you probably have to use this->pos_ to get to it.
It's not about clarity, it's the fact that MenuBase already manages a menu position for you, so I think you should use it, as it makes for less code.

If you're not going to use it, you probably want SteppedMode again...

NoSloppy replied:
> use this->pos_

Would you be ok with a helper function to help readability? In order to account for the offset, things can get kind of unruly like `this->pos_ = (this->pos_ + size() - 1) % size();`
I could do something like
```cpp
  int current_blade() const {
    return (this->pos_ % size()) + 1;
  }
  ```
and then clean things up by just calling that?
Then it's still using pos_ as requested.

profezzorn replied:
this->pos_ should always be between 0 and size() -1, so the % doesn't help.
A helper function  can be nice though, as it gives a better name than "pos_"

NoSloppy replied:
but i tried that and it cycled through blade 0-3 not 1-4.

NoSloppy replied:
> MenuBase already manages a menu position for you, so I think you should use it, as it makes for less code.

I'm trying to abide by this, but I don't see how to "do" anything when prev() or next() happens if I can't override them, and if they are overridden, then it's just new, replacement code anyway, in which case enforcing using pos_ seems weird.
If not using pos_ enforcing inheriting from SteppedMode is weird because I don't need the methods it uses.
I'm in a catch 22 it seems.

profezzorn replied:
What do you want to do in prev() / next()?

> If not using pos_ enforcing inheriting from SteppedMode is weird because I don't need the methods it uses.

MenuBase is just SteppedMode, but with an added size and position, if you don't want the size and position, then why inherit from MenuBase?

If you don't need the methods from SteppedMode, then maybe you should inherit from SteppedModeBase?

> I'm in a catch 22 it seems.

I don't understand the catch 22 here.

NoSloppy replied:
> then maybe you should inherit from SteppedModeBase?

oh. I hadn't seen you just made this.

If not inheriting from MenuBase, the modolo is needed, right? 


> What do you want to do in prev() / next()?

Stop the current blade highlighting, increment/decrement the blade #, start highlighting the new blade.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1134,317 @@ EFFECT(volmax);     // for maximum volume reached
+  }
+
+  void exit() override {
+    highlighted_blade_.Stop(current_blade_);
```
Better to do this in mode_deactivate()

NoSloppy replied:
but mode_deactivate is only when leaving the mode, right?
This is here so that when going to the next or prev blade, the prior one shuts off.

profezzorn replied:
exit() isn't called when going to the next or previous option

NoSloppy replied:
ok it is actually deactivating though, so I added a mode_deactivate() and moved the stopping there.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1134,317 @@ EFFECT(volmax);     // for maximum volume reached
+    // Deactivate current mode and push to LED length adjustment mode for the selected blade
+    selected_blade_ = current_blade_;
+    highlighted_blade_.Stop(current_blade_);
+    SPEC::SteppedMode::mode_deactivate();
```
There is no need to call mode_deactivate explicitly, pop_mode takes care of that.

NoSloppy replied:
cool. removed.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1134,317 @@ EFFECT(volmax);     // for maximum volume reached
+#endif
+
+#ifdef BC_ENABLE_OS_MENU
+#define MENU_SPEC_TEMPLATE DefaultMenuSpec
```
MENU_SPEC_TEMPLATE is meant to be controlled by the config file.
How about just doing:

#ifndef MENU_SPEC_TEMPLATE
#define MENU_SPEC_TEMPLATE BCMenuSpec
#endif


?

NoSloppy replied:
ok let me try to explain.
This was trying to workaround guarded code in prop_base. Maybe that code isn't expecting custom specs, but I needed sound library polled (now init() ) but it is inside `ifdef MENU_SPEC_TEMPLATE`.
I thought to just satisfy that by defining MENU_SPEC_TEMPLATE to BCMenuSpec at the beginning of my prop file, but then it complains that TopMenu isn't in the custom spec due to how EnterMenu() gets set.

So the workaround was to first do `#define MENU_SPEC_TEMPLATE DefaultMenuSpec` at the top of my prop to satisfy the prop_base stuff, then undef it, then re-define it according to BC_ENABLE_OS_MENU or not.

Hence, that's what you see in the code you commented on above.

Now I'm gathering I could just do `SoundLibrary::init()` in my prop somehow instead of all that runaround?
What would be the syntax for that, because I tried using the same line from prop_base `MKSPEC<MENU_SPEC_TEMPLATE>::SoundLibrary::init();` 
which says not in scope.
Then I tried to preface it with just `#define MENU_SPEC_TEMPLATE`
but then it says `wrong number of template arguments (0, should be 1)`

profezzorn replied:
> So the workaround was to first do #define MENU_SPEC_TEMPLATE DefaultMenuSpec at the top of my prop to  satisfy the prop_base stuff, then undef it, then re-define it according to BC_ENABLE_OS_MENU or not.

This doesn't work at all.
enterMenu() will do the wrong thing, and init() will be called on the wrong spec.

I think the way to solve this is to not set MENU_SPEC_TEMPLATE at all if your menus are being used.
You'll need to add an init call, like:

#ifndef MENU_SPEC_TEMPLATE
   MKSPEC<BcMenuSpec>::SoundLibrary::init();
#endif

This assumes that your menus are mutually exclusive with the prop_base ones, is that the case?

Basically, if the user defines MENU_SPEC_TEMPLATE it would disable your menues and use some other ones.

Setting MENUS_SPEC_TEMPLATE to BcMenuTemplate doesn't really work, since there is no TopMenu...

NoSloppy replied:
ok. getting an error with that, error: invalid use of template-name 'MKSPEC' without an argument list
 1194 | MKSPEC::SoundLibrary::init();
 
 but once it's right, could it be like this so that BC_ENABLE_OS_MENU remains the thing to choose with?
```

#ifndef PROPS_SABER_BC_BUTTONS_H
#define PROPS_SABER_BC_BUTTONS_H

// #define MENU_SPEC_TEMPLATE DefaultMenuSpec
#ifdef BC_ENABLE_OS_MENU
#define MENU_SPEC_TEMPLATE DefaultMenuSpec
#else
MKSPEC::SoundLibrary::init();
#endif

#include "prop_base.h"
```

profezzorn replied:
Grr, it thinks I was writing HTML and hid it, let's see if this works better:

```cpp
MKSPEC<BcMenuSpec>::SoundLibrary::init();
```
The init() call needs to go in the prop constructor, or Setup() though, since it's a regular function call that needs to be executed.

> could it be like this so that BC_ENABLE_OS_MENU remains the thing to choose with?

I'm not sure I understand why you would want that?

> #define MENU_SPEC_TEMPLATE DefaultMenuSpec

Doing this makes it impossible for the user to select a different menu spec.
Currently there is only one, but that is likely to be short-lived.
Also, I want people to be able to put modified SPECs in their config files if they want to.

> Will the BCMenuSpec be known/used then without setting MENU_SPEC_TEMPLATE to it?

If your prop calls it, then it will be known/used, right?

> Still fuzzy on all this.

Well, the way I envisioned it is that prop makers don't really have to worry too much about the menus. My thinking was that there is one top-level menu, and all the prop has to do is to provide a key combination for entering the menu, the rest is up to the user, as they can then select which menu system and which sub-menus should be available by selecting a spec and which top menu to use.

However, I also want this to be as customizable as humanly possible, and my idea isn't the only way to do things. So if prop makers like you have ideas for how the menus should work, I want the system to be able to support that.  Since it's early days and nothing is set in stone yet, I'm more than happy to rewrite some stuff to make things work well for you.

The key piece that I feel that we're struggling with is separating "it's hard to make it work" from "i'm trying to do X". I'm more than happy to help with both, but it's much easier to do it well if I have a good understanding of what you are trying to achieve.

NoSloppy replied:
> prop makers don't really have to worry too much about the menus. My thinking was that there is one top-level menu, and all the prop has to do is to provide a key combination for entering the menu, the rest is up to the user, as they can then select which menu system and which sub-menus should be available by selecting a spec and which top menu to use.

I guess I was trying to do the opposite. Make it a one define choice for the user to avoid swimming in a sea of too many options. Choose A and use the BC defaults, or B and use the full OS menu from top level down.
I don't want to restrict choice B though. I think it should be open for customization in the user config, just defaulted to the DefaultMenuSpec TopMenu.

Assuming there's going to be an easy-to-follow cheat sheet of sorts, as I'm not sure how many average users will be customising their menu experience on their own, and if they do want to, they'd want it as simple as possible. I think "MENU_SPEC_TEMPLATE" alone might be enough to scare off some people who would just rather choose A or B and be on their way. 
I, for one, can obviously appreciate all the customization options as a user and don't mind tailoring which menus to use and such. The flexibility is an awesome thing. But not everybody is like that.
So I'd like to find a middle ground, where using the BC prop gives you a default of custom menus (made for me because it's mine and that's how I'd like it), and if you want to customize, go for it!

NoSloppy replied:
All that said, I think I got it. I updated the code, killed BC_ENABLE_OS_MENU, and updated the in-file documentation. 
I also did not know Setup() was a thing so that's great.

profezzorn reviewed (commented):
Read some of it, here is the first batch of comments.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1134,321 @@ EFFECT(volmax);     // for maximum volume reached
+    initial_volume_ = dynamic_mixer.get_volume();
+    initial_percentage_ = round((initial_volume_ / max_volume_) * 10) * 10;
+    last_volume_ = initial_volume_;
+    current_volume_ = initial_volume_;
```
Isn't current_volume_ always the same as dynamic_mixer.get_volume()? (If so, why have it?)

NoSloppy replied:
I figured using the variable would reduce calls to dynamic_mixer.get_volume(), as well as make it a bit more readable? like in comparisons of `if (current_volume_ < max_volume_)`

NoSloppy replied:
Actually, it's last_volume_ that's the superfluous thing. It was used in a previous iteration, but could go away now. Removing.

profezzorn replied:
but isn't current_volume_ *also* superfluous?

profezzorn replied:
> I figured using the variable would reduce calls to dynamic_mixer.get_volume()

Pretty sure the compiler will figure that out and make it equally efficient as accessing a variable.

> as well as make it a bit more readable? 

It is true that the code can become more readable, but the way to do that is to use *local* variables. (variables which only exists within one function.) When you persist the variable between functions, the chance of making a mistake somewhere so that current_volume_ and dynamic_mixer.get_volume() are *different* becomes significant, and checking for those sort of problems adds to the time it takes to read the code.

Generally speaking, avoiding storing the same thing in two places is one of the fundamental principles of good code.

NoSloppy replied:
ok. moved current_volume_ inside functions and only where useful.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1134,321 @@ EFFECT(volmax);     // for maximum volume reached
+#endif  // BC_DUAL_BLADES
+
+
+template<class SPEC>
```
Besides handling different blades, is there any difference between this and the mode in modes/ ?

NoSloppy replied:
Just the user experience mainly. I realize I probably don't need to override all the methods from the base if it inherits from ChangeBladeLengthBlade1, only select() and upadte() really, but I have a real hard time making that inheritance happen. Something about the naming, or circular dependency, or having the typedef use the same name etc... I got to a point it compiled and seemed to work, but ShowLengthStyle won't light up an LED.
I just left it as an independent thing once it worked.
If you prefer me to change it, great, but I'll need some coaching on what needs to be changed.

profezzorn replied:
> Just the user experience mainly.

That can mean anything though.

> If you prefer me to change it, great, but I'll need some coaching on what needs to be changed.

I don't mind the coaching, but I have to know what the intended differences are first.

NoSloppy replied:
> I don't mind the coaching, but I have to know what the intended differences are first.



ok I'll explain instead of trying to figure it all out on my own (trying to save you hassle).
The only differences would be
steps_per_revolution 40.
Use SFX_volup instead of ccchange, plus delay the announce by 400ms.
SaySave on exit instead of SaySelect.

As far as what's confusing/not working when I try it: (this is my brain)
I see that BCChangeBladeLengthMode is mostly redundant of the already existing ChangeBladeLengthBlade1 and realize that's not the right way to do it likely.
I wanted to change it to just inherit what it could and only change what I wanted to customise. That's how it works, right?

Well I originally had the BCChangeBladeLengthMode using a "BC" version alias in the typedef as `typedef BCChangeBladeLengthMode<SPEC> BCChangeBladeLengthMode;`, but testing yielded no LED lighting up. So looking at ShowLengthStyle, I saw that it's looking for 'SPEC::ChangeBladeLengthMode>()->getLength();'.
(this is where I probably messed up) So instead of making another thing, and duplicating ShowLengthStyle to a new version like BCShowLengthStyle just so that it could use BCChangeBladeLengthMode... I just typedef-ed the BCChangeBladeLengthMode to the same alias name as the default ChangeBladeLengthBlade1 template uses, ChangeBladeLengthMode, thinking that would work since I wasn't using the actual ChangeBladeLengthBlade1 template in my spec, so the alias shouldn't be in conflict or anything.
Well that didn't work, I don't remember if it was a compiler error or the LED still wasn't lighting, but then I said let's do this right and use inheritance from the default version.
So I restored the BCversion alias in the typedef, and tried more things and it still didn't work.
At that point, i just left it as inheriting from MenuBse, fully redundantly overrode all the methods, and that's where it stands because it works.

Bottom line is I really only need to override steps_per_revolution, select(), and update().

profezzorn replied:
I indent to create a define which lets you specify which blades are "removable", and then change the default menu so that you can change the length of all the removable blades. When I do that, hopefully I can also make all of this simpler for you to use and override.

It's not always true that inheriting and overriding is the right thing to do. It kind of depends on how many things you want to change. If you want to change one thing, then inherit-and-override is almost certainly the right way to go, but if you want to change enough things that you have to override almost everything, then it might be better and easier to just copy-paste-modify, that is normal and reasonable.

I don't really have any major problems with this way of implementing it, other than that I kind of hope that I can change the existing implementation to make it easier for you to do what you want, and maybe then inherit-and-override will be the right way to solve things again.

NoSloppy replied:
> I intend to create a define which lets you specify which blades are "removable", and then change the default menu so that you can change the length of all the removable blades. When I do that, hopefully I can also make all of this simpler for you to use and override.

Sounds good. Maybe it'll be what gets used here instead of NUM_BLADES.
So for now this can remain as-is? 
The only thing I'm not thrilled about is how it "cheats" to make use of ShowLengthStyle. It only works I think because I've aliased BCChangeBladeLengthMode as ChangeBladeLengthMode (the same thing the default spec uses.)
I think it really should be aliased as like BCChangeBladeLengthMenu instead, but then would need to also make a BCShowLengthStyle version too.
 Should I do that?

NoSloppy replied:
This all has been resolved by inheriting from the SPEC directly instead of using the typedef-ed alias name.

NoSloppy replied:
> > Bottom line is I really only need to override steps_per_revolution, select(), and update().
> 
> Can we try doing that? It seems like it should be nicer and smaller that way. (And I think other people may end up copying from your code, so I want it to be as neat at possible.)

Done.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1134,321 @@ EFFECT(volmax);     // for maximum volume reached
+  typedef BCChangeBladeLengthMode<SPEC> ChangeBladeLengthMode;
+#endif  // DYNAMIC_BLADE_LENGTH
+  typedef mode::SelectCancelMode SelectCancelMode;
+  typedef mode::SteppedMode<SPEC> SteppedMode;
```
You'll need to add SteppedModeBase now...

NoSloppy replied:
ok yes found that, done just not updated yet.

NoSloppy replied:
done

NoSloppy replied:
Am I allowed to resolve small, clear cut-and-dry items like this?

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1136,495 @@ EFFECT(volmax);     // for maximum volume reached
+
+  void prev() override {
+    highlighted_blade_.Stop(this->pos_ + 1);
+    if (this->pos_ == 0) {
```
1482..1486 can just be SPEC::MenuBase::prev() I think.

NoSloppy replied:
oh we can do that?! Call an overridden method from within the one you're overriding it with ?
Wow well that certainly makes life easier. Great. How does it look now then? (tested working).
Hopefully that wraps up this BCSelectBlade thing. 
Lots of current conversations, don't want to get lost. I'm going to stick with one at a time until good.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1136,495 @@ EFFECT(volmax);     // for maximum volume reached
+    mode::getSL<SPEC>()->SaySelectBlade();
+    ShowColorStyle::SetColor(Color16(65535, 65535, 65535));
+    SPEC::SteppedMode::mode_activate(onreturn);
+    highlighted_blade_.Start(this->pos_ + 1);
```
I liked your idea of a helper function: int current_blade() { return this->pos_ + 1; }

NoSloppy replied:
done

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1136,297 @@ EFFECT(volmax);     // for maximum volume reached
+  }
+
+  void next() override {
+    highlighted_blade_.Stop(current_blade());
```
An alternate way to do this (which may or may not be better, I'm not actually sure) would be to save which blade is highlighted, and then do something like:

if (current_blade() != highlighted_blade_num_) {
  highlighted_blade_.Stop(highlighted_blade_num_);
  highlighted_blade_num_ = current_blade();
  highlighted_blade_.Stop(highlighted_blade_num_);
}
 
This code could be in update(), which means you would only need to do it once.
(Note the current code is fine, this is an optional idea.)

NoSloppy replied:
ok to leave for now? Maybe if a non-prop version develops in the future and uses that way, then the prop version could adjust to follow suit?

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1136,297 @@ EFFECT(volmax);     // for maximum volume reached
+    SPEC::MenuBase::prev();
+    highlighted_blade_.Start(current_blade());
+    PVLOG_NORMAL << "** Highlighting blade: " << current_blade() << "\n";
+    say();
```
I don't think this call to say() is needed.
Since you're inheriting from MenuBase, update() will be called after this function, which will call fadeout(), which will essentially cancel anything this call is doing. say() will be called after ccchange.wav  / beep is played.

NoSloppy replied:
Indeed. Confirmed works fine without this. Removed.
I've also moved SayEditBladeLength() to the button event call instead, and removed "Blade" before the number.
It was too much, especially when exiting to have to hear each time "Edit blade length, select blade, blade 1".

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1136,297 @@ EFFECT(volmax);     // for maximum volume reached
+    SPEC::MenuBase::next();
+    highlighted_blade_.Start(current_blade());
+    PVLOG_NORMAL << "** Highlighting blade: " << current_blade() << "\n";
+    say();
```
see comment below

NoSloppy replied:
done

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1136,297 @@ EFFECT(volmax);     // for maximum volume reached
+};
+
+#ifdef DYNAMIC_BLADE_LENGTH
+int selected_blade_;
```
Would be nice to use menu_current_blade instead.

NoSloppy commented:
Ok I've made significant changes on all the menu stuff as per our discussions above, as well as some button tweaks (not using a toggle for main and second blades on/off anymore).
Should I just drop the new file in overwriting the current version here? I'm just saying most of the comments above might get lost? (although I think they've all been addressed).

profezzorn commented:
Yes, please drop the new file.
Every now and then a clean slate is needed.

NoSloppy commented:
This comment is here as a reminder for now:
There are a few things in here that might already be handled by prop base or elsewhere and don't need to be specifically addressed here in the prop. When we get up to them, I'd like to have clarification.
Like I see in DetectTwist(), there's reference to TWIST_LEFT : TWIST_RIGHT
and also thrust stuff, like SHAKE_FWD and SHAKE_REW a.k.a "Yank".
Those might be handy for gestures I tried to make work in here.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -383,34 +1158,253 @@ EFFECT(voldown);    // for decrease volume
 EFFECT(volmax);     // for maximum volume reached
 #ifdef ENABLE_FASTON
-#warning The faston.wav sound will be replaced with fastout.wav. If you have a good reason to keep faston.wav as is, please post at https://crucible.hubbe.net/
+// #warning The faston.wav sound will be replaced with fastout.wav. If you have a good reason to keep faston.wav as is, please post at https://crucible.hubbe.net/
```
Commented out warning does nothing?
remove or uncomment?

NoSloppy replied:
Uncommented. done.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -383,34 +1158,253 @@ EFFECT(voldown);    // for decrease volume
-EFFECT(monosfx);    // for Monophonically played sounds (iceblade, seismic charge etc...)
-EFFECT(swap);       // for standalone triggering EffectSequence<>
+EFFECT(tr);         // for EFFECT_TRANSITION_SOUND, use with User Effects.
+EFFECT2(trloop, trloop);  // have YET to play with this
```
If it's not used, why add it?

NoSloppy replied:
Removed.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -383,34 +1158,253 @@ EFFECT(voldown);    // for decrease volume
 EFFECT(mute);       // Notification before muted ignition to avoid confusion.
+EFFECT(mzoom);      // for Spam Blast enter/exit
+
+#ifndef MENU_SPEC_TEMPLATE
```
I think you should remove the #ifndef
The code will be read and checked for errors, but not actually used, which tends to be beneficial.

NoSloppy replied:
ah ok. That might apply later below with other guards as well. Done.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -383,34 +1158,253 @@ EFFECT(voldown);    // for decrease volume
+  void next() override {
+    int current_volume_ = dynamic_mixer.get_volume();
+    if (current_volume_ < max_volume_) {
+      current_volume_ += max_volume_ * 0.10;
```
current_volume_ can end up higher than max_volume_ here.
Maybe add:
if (current_volume_ > max_volume_) current_volume_ = max_volume_;

(Or some std::min call)

NoSloppy replied:
Perfect, thanks. Added this here, and also limited the min in `prev()`

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -383,34 +1158,253 @@ EFFECT(voldown);    // for decrease volume
+    percentage_ = round((volume / max_volume_) * 10) * 10;
+    SaberBase::DoEffect(EFFECT_VOLUME_LEVEL, 0);
+    // Overridden to substitute the tick sound
+    this->say_time_ = Cyclint<uint32_t>(millis()) + 1;
```
I think you can skip setting say_time_ since there is no say() function in this class.

NoSloppy replied:
Indeed. Done.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -383,34 +1158,253 @@ EFFECT(voldown);    // for decrease volume
+  }
+  void mode_activate(bool onreturn) override {
+    mode::getSL<SPEC>()->SaySelectBlade();
+    ShowColorStyle::SetColor(Color16(65535, 65535, 65535));
```
There doesn't seem to be a need to use ShowColorStyle here.
Why not just use "White" in the style template on line 1289?

NoSloppy replied:
That works for me, thanks.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -383,34 +1158,253 @@ EFFECT(voldown);    // for decrease volume
+    pushMode<typename SPEC::ChangeBladeLengthMode>();
+  }
+
+  void exit() override {
```
An overridden function that just calls the inherited base function can be omitted.

NoSloppy replied:
Removed.

NoSloppy commented on props/saber_BC_buttons.h:
```diff
@@ -308,6 +1033,12 @@ Turn OFF without postoff - Turn OFF while pointing up.
+#include "../modes/select_cancel_mode.h"
+#include "../modes/settings_menues.h"
+
+#ifndef MIN_SOUND_LIBRARY_VERSION
```
got it. removing.

NoSloppy commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1478,1227 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+    thrust_location = EffectLocation(0, thrusting_blade_);
+  }
+
+  bool IsStab(const Vec3& diff) override {
```
This seems to get called a bunch, and it's fluctuating a lot with what it sees.
It's not really trustworthy to choose the correct blade, sort of a 50/50 chance in my testing.
Also, it seems for some reason that this as well as Thrust calculations work better when the target blade is angled slightly down?
Maybe there's some filtering that could be applied to smooth out the readings so that a better decision could be made?

profezzorn replied:
When this is used for stab, only the return value at the time that the clash is triggered actually counts. All the other calculations don't do anything.

However, since you are changing forward_stab_ inside the function, and this function may be called a bunch more times before the event is processed, you get bad values.

At least I think that's what's going on.

Breaking out IsStab() might not have been the right way to do this, because we don't actually have all the information we need in this function now...

NoSloppy replied:
> When this is used for stab, only the return value at the time that the clash is triggered actually counts. All the other calculations don't do anything.
> 
> However, since you are changing forward_stab_ inside the function, and this function may be called a bunch more times before the event is processed, you get bad values.
> 
> At least I think that's what's going on.
> 
> Breaking out IsStab() might not have been the right way to do this, because we don't actually have all the information we need in this function now...

This was correct. So I reverted IsStab() and restored DoAccel() completely in prop_base, see https://github.com/profezzorn/ProffieOS/pull/692.
Then I just made a full DoAccel() override in my prop, and added the foward_stab_ detection in the right place.
See latest update.
Works in testing.

NoSloppy commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1478,1227 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
     if (scroll_presets_timer_.timerCheck() && scroll_presets_) {
         SaberBase::DoEffect(EFFECT_NEWFONT, 0);
     }
+    if (quote_delay_timer_.timerCheck()) {
```
Meaning since now there's nothing executed, just do the check. Got it.
Done.

NoSloppy commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1478,1227 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+  float GetSoundLengthNow(Effect* effect) {
+    RefPtr<BufferedWavPlayer> tmp = GetWavPlayerPlaying(effect);
+    if (tmp) {
+      tmp->UpdateSaberBaseSoundInfo();
```
removed, see above

NoSloppy commented:
> When this is used for stab, only the return value at the time that the clash is triggered actually counts. All the other calculations don't do anything.
> However, since you are changing forward_stab_ inside the function, and this function may be called a bunch more times before the event is processed, you get bad values.
>At least I think that's what's going on.
>Breaking out IsStab() might not have been the right way to do this, because we don't actually have all the information we need in this function now...

This was correct. So I reverted IsStab() and restored DoAccel() completely in prop_base, see #692.
Then I just make a DoAccel() override in my prop, added the detection in the right place.
See latest update. 
Works in testing.

NoSloppy commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1478,1227 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
-        battle_mode_ = true;
-#endif
-        last_twist_ = millis();
+  void DoChangePreset() {
```
Well it's doing previous, next, or first preset, so I had thought DoChangePreset made sense.
Would you prefer DoChoosePreset? 
I'm open to suggestions of course.

NoSloppy replied:
How about this?

profezzorn replied:
> How about this?

What changed?

> I could also just add "// Previous, next, or first preset, depending on blade angle"

That seems like a useful comment, please add.

NoSloppy replied:
Sorry, that was a "bump" to bring attention to an unresolved comment.
I meant to ask if any of those alternate names were better, or if the comment is enough.
I've added the comment for now, please let me know otherwise.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1491,1257 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+    if (hybrid_font.PlayPolyphonic(&SFX_mute)) {
+      mute_mainBlade_timer_.trigger(SaberBase::sound_length * 1000);
+    } else {
+      if (SetMute(true)) {
```
This code is duplicated here and in loop().
Duplicated code is generally a Bad Thing (tm).
It seems easy to avoid here though, as you could just always call mute_mainBlade_timer_.trigger().
If SFX_mute doesn't exist, then just call it with a delay of zero, right?

NoSloppy replied:
oh that's great, yes.
Changed for all 3 mute timers (all, main, and second)

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1491,1257 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
 
-// Enter / Exit Volume MENU
-#ifndef NO_VOLUME_MENU
+      bool stab = (diff.x < -2.0 * sqrtf(diff.y * diff.y + diff.z * diff.z) ||
```
X < -A || X > A  => fabs(X) > A

NoSloppy replied:
Very nice. Done.

NoSloppy commented on props/saber_BC_buttons.h:
```diff
@@ -2,82 +2,123 @@
-                          blade effect too by using EFFECT_USER2 in a TransitionEffectL.
+User Effects: (a.k.a. Fett263 "Special Abilities")
+- Can be built into blade style code to trigger anything that takes an EFFECT_XXXXX argument.
+EFFECT_USER1
```
Better now?
I rearranged it a bit and added clarification that they equate to Fett263 "Special Abilites", but I am *not* calling them that.
Any other difference is really just the gesture, which is detailed further down with the rest of the controls.

NoSloppy commented on props/saber_BC_buttons.h:
```diff
@@ -2,82 +2,123 @@
+  - OS System Menu system available with a #define.
+* Deprecated features:
+- No Volume UP/DOWN with buttons. Use rotary control instead (see above).
+                          QuickMinVolume and QuickMaxVolume are still available.
```
repaired.

NoSloppy commented on props/saber_BC_buttons.h:
```diff
@@ -2,82 +2,123 @@
+- * NEW for OS8 * 
+  - Dual blade independent ignition and retraction control with a single Proffieboard. (Such as for a staff saber)
+  - Scroll Presets mode.
+  - Quotes play  with blade ON or OFF, and will not overlap / interrupt one another.
```
removed.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -364,14 +1120,33 @@ Turn OFF without postoff - Turn OFF while pointing up.
+
 #if defined(NO_BLADE_NO_GEST_ONOFF) && !defined(BLADE_DETECT_PIN)
-#error Using NO_BLADE_NO_GEST_ONOFF requires a BLADE_DETECT_PIN to be defined 
+#error Using NO_BLADE_NO_GEST_ONOFF requires a BLADE_DETECT_PIN to be defined
```
This might be a TODO for later, but doesn't NO_BLADE_NO_GEST_ONOFF work with using continuous blade detect and NO_BLADE_ID_RANGE?

NoSloppy replied:
It could. I added BLADE_ID_SCAN_MILLIS to the error, is it ok to do the 2 small changes I made to prop_base?
- Unguarding this global variable
```cpp
#ifdef BLADE_DETECT_PIN
  bool blade_detected_ = false;
#endif
```
- and adding this to ScanBladeIdNow()
`      blade_detected_ = (current_config->ohm < NO_BLADE);`

profezzorn replied:
replace blade_detected_ with blade_present() in your prop instead.

NoSloppy replied:
omg I didn't even see that and I was right above it too.
Thanks, ok done.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1161,254 @@ EFFECT(volmax);     // for maximum volume reached
+    // If only one blade, jump right to ChangeBladeLengthMode
+    // If just adjusted and returning here via popMode, just exit.
+    if (size() == 1) {
+      if (!single_blade_adjusted_) {
```
I think there are at least three better ways to do this....

1. You can use the "onreturn" variable to detect that we got here by popping rather than by pushing
2. you could pop this mode before pushing the ChangeBladeLengthMode so it's not on the mode stack at all anymore.
3. you could use #if NUM_BLADES == 1 and where this mode is used and use ChangeBladeLengthMode directly if that is the case.

NoSloppy replied:
oh wow `onreturn` has been there the whole time huh.
I was doing things in Volume menu like "first_time_through_" or whatever and this is readily available to do that exact job. So nice, thanks.
Adjusted.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -425,6 +1423,26 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+#endif  // DYNAMIC_BLADE_LENGTH
+  }
+
+  void EnterVolumeMenu() {
```
EnterVolumeMenu can be implemented (and be useful) even if MENU_SPEC_TEMPLATE is defined.

NoSloppy replied:
I suppose so, as an additional / alternate / shortcut version sort of thing?
Unguarded.

profezzorn replied:
I wonder, if MENU_SPEC_TEMPLATE is defined, wouldn't it make more sense to use that volume menu:

   pushMode<MKSPEC<MENU_SPEC_TEMPLATE>::ChangeVolumeMode>()

?

NoSloppy replied:
Oh I see, shortcut to the OS menu version. 
Yeah, ok. I'll add an announce too so it's known what happened.
Done.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1495,1272 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
-      } else {
+// EVENT_THRUST
+    if (mss.y * mss.y + mss.z * mss.z < 20.0 &&  // slightly more forgiving than 16.0?
+       (mss.x > 14 || mss.x < -14) &&  // Check for both positive and negative x-motion
```
mss.x > 14 || mss.x < -14   ->  fabs(mss.x) > 14

NoSloppy replied:
Yes. Done.

profezzorn replied:
Doesn't look done?

NoSloppy replied:
it is now.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,31 +1161,254 @@ EFFECT(volmax);     // for maximum volume reached
+      return false;  // Timer is still running
+  }
+
+  bool isTimerRunning() const {
```
I think this function should call timerCheck() (and not be const)

NoSloppy replied:
That is more accurate. How about losing the triggered_ variable too.
See update.

NoSloppy replied:
Actually that doesn't work, I'm getting premature resets.
  This works instead though.
  ```cpp
  bool isTimerRunning() {
    return trigger_time_ != 0 && (millis() - trigger_time_ < duration_);
  }
  ```

profezzorn replied:
What I was imagining was:
bool isTimerRunning() {
  timerCheck();
 return triggered_;
}

NoSloppy replied:
ohhh. ok yeah that works. 
Updated.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1495,1272 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+    if (twist_timer_.timerCheck() && saved_twist_) {
+      DoSavedTwist();
+    }
+    quote_delay_timer_.timerCheck();
```
If isTimerRunning calls timerCheck(), then you don't need to call it here, right?

NoSloppy replied:
This line doesn't need to be here. 
Removed.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1495,1272 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+  void DetectMenuTurn() {
+    float a = fusor.angle2() - current_menu_angle_;
+    if (isPointingUp()) return;
+    // Keep the rotational angle within range of -180 to 180.
```
Seems confusing to use degrees in the comment and radians in the code...

NoSloppy replied:
It was a note-to-self really when realizing what it was doing. 
No harm in keeping it, right? But I added "in terms of radians".

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1495,1272 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
     if (scroll_presets_timer_.timerCheck() && scroll_presets_) {
         SaberBase::DoEffect(EFFECT_NEWFONT, 0);
     }
+    if (twist_timer_.timerCheck() && saved_twist_) {
```
I'm not sure why you have to call timerCheck AND saved_twist_ here.
How can timerCheck return true if saved_twist_ is false?

NoSloppy replied:
ah true. Timer wouldn't be running if no saved_twist_ is set, so there's no need to check both .
Done.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1495,1272 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+    if (a > M_PI) a-= M_PI * 2;
+    if (a < -M_PI) a+= M_PI * 2;
+
+    if (a < -M_PI / 6) {
```
Is this for menus, or is it for "twist" events? It's quite different from the twist detection in the base prop I think.

NoSloppy replied:
This is used for USER_EFFECT twists, doing EVENT_TWIST_LEFT and EVENT_TWIST_RIGHT.
Separate sensitivity from menus.

profezzorn replied:
Ok, so this is supposed to be used while on?
Since it only checks rotation angle, I would assume that this gesture would occur very often "by accident" ?

NoSloppy replied:
They require a hold of a button, so it's not easy to do accidentally like
`case EVENTID(BUTTON_NONE, EVENT_TWIST_LEFT, MODE_ON | BUTTON_POWER):`

profezzorn replied:
Is it always the power button?
I'm thinking that it might be a good idea to reset current_menu_angle_ (which should probably be renamed) whenever the button *isn't* pressed... That way the angle required will be consistent.

NoSloppy replied:
> Is it always the power button?
> I'm thinking that it might be a good idea to reset current_menu_angle_ (which should probably be renamed) whenever the button _isn't_ pressed... That way the angle required will be consistent.

No, the AUX button when 2 buttons also does them.
Could you explain this? I was just playing around with it and was questioning the "start" position when editing, but settled on how it is right now as working-as-intended.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1495,1272 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+    if (!saved_twist_) {
+      // Save the current twist and start the timer if no twist is saved
+      saved_twist_ = event;
+      twist_timer_.trigger(DOUBLE_TWIST_TIMEOUT);
```
What is a "double twist" ?

NoSloppy replied:
I know, that's a bad choice. But "TWIST_TIMEOUT" seems unclear though, even though it's not specified LEFT or RIGHT.

It's just "regular twist", the 2-direction gesture.
Is "regular twist" better? 
"bi-directional twist"?
"normal twist"
I'll go with REGULAR_TWIST_TIMEOUT. Let me know otherwise.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -386,36 +1161,254 @@ EFFECT(volmax);     // for maximum volume reached
+  uint16_t size() override { return NUM_BLADES; }
+
+  void mode_activate(bool onreturn) override {
+    // If only one blade, and returning here via popMode, just exit.
```
if there is only one blade, why enter this menu at all?

NoSloppy replied:
I guess I still don't get pushing and popping so well.
When finishing BCChangeBladeLengthBlade1, it pops back to BCSelectBladeMode, even if we pushed directly to BCChangeBladeLengthBlade1 because NUM_BLADES ==1.
```cpp
#if NUM_BUTTONS > 1
        pushMode<MKSPEC<BCMenuSpec>::BCSelectBladeMenu>();
#else
        pushMode<MKSPEC<BCMenuSpec>::ChangeBladeLengthMode>();
#endif
```
If you mean to exit differently based on NUM_BLADES, I'm going to need a hint.

NoSloppy replied:
Wait. No it doesn't pop back to BCSelectBladeMode. This whole part can go away.
Done.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1493,1269 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
-      case EVENTID(BUTTON_AUX, EVENT_PRESSED, MODE_ON):
-      case EVENTID(BUTTON_AUX2, EVENT_PRESSED, MODE_ON):
-      return true;
+  void DoChangePresetFast() {
```
This could also use a comment.

NoSloppy replied:
done.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1493,1269 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+  // Handle lockup or drag for dual blades when both blades are on
+  void DualBladesLockupOrDrag() {
+    if (fusor.angle1() < -M_PI / 4) {
+      // Main blade is pointing down, "return_opposite_" and apply drag.
```
What does "return_opposite_" mean?

NoSloppy replied:
Gone.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1493,1269 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+  bool isSecondBladeOn() { return SaberBase::OnBlades()[BC_SECOND_BLADE]; }
+
+  // Get blade above horizon line to apply effect to
+  BladeSet GetBladeAboveHorizon(bool return_opposite = false) {
```
return_opposite is a makes a travesty of the name of this function.
Why not just have a GetBladeBelowHorizon() instead?

NoSloppy replied:
This whole double-negative thing had me crazy and this worked so I left it. Agreed, return_opposite is not great.
Looking now, there's no reason we can't just do drag to the right blade directly in DualBladesLockupOrDrag().
Then GetBladeAboveHorizon() can be just what it says to handle clash, blast, and lockup.
Tested and works the majority of the time.
I think blade wiggle/vibrations messes with accel sometimes so the "wrong" blade shows impact effects, but it's mostly acceptable I think?
Edits done.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1492,1266 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
-#endif  // BC_TWIST_ON
+  void OnDemandBatteryLevel() {
+    if (scroll_presets_) return;
+      PVLOG_NORMAL << "Battery Voltage: " << battery_monitor.battery() << "\n";
```
weird indentation.

NoSloppy replied:
fixed.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1492,1266 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+      PVLOG_NORMAL << "Battery Voltage: " << battery_monitor.battery() << "\n";
+      speaking_ = true;
+      SaberBase::DoEffect(EFFECT_BATTERY_LEVEL, 0);
+      return;
```
redundant return

NoSloppy replied:
fixed.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1492,1266 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+      speaking_ = true;
+      SaberBase::DoEffect(EFFECT_BATTERY_LEVEL, 0);
+    }
+    return;
```
redundant return

NoSloppy replied:
fixed.

profezzorn replied:
doesn't look fixed...

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1492,1266 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
-        battle_mode_ = true;
-#endif
+  #ifdef BC_DUAL_BLADES
+      if (isMainBladeOn() && !isSecondBladeOn()) {
```
weird indentation

NoSloppy replied:
The preprocessor lines weren't left justified, others look ok though, no?

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1492,1266 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
-      case EVENTID(BUTTON_AUX, EVENT_FIRST_CLICK_LONG, MODE_OFF):
+  void ToggleSequentialQuote() {
+    if (scroll_presets_ || spam_blast_) return;
+    if (spam_blast_) return;
```
redundant if()

NoSloppy replied:
oops. fixed.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1492,1266 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+
+  void ToggleBattleMode() {
+#ifndef BC_NO_BM
+  battle_mode_ = !battle_mode_;
```
weird indentation

NoSloppy replied:
fixed.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1492,1266 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+
+  // Get blade location to apply effect to
+  BladeSet GetBladeAboveHorizon() {
+    return accel_.x > 0 ? BC_MAIN_BLADE_SET : BC_SECOND_BLADE_SET;
```
Actually, it's probably better to use fusor.down().x instead of accel_

profezzorn replied:
or fusor.angle1()

NoSloppy replied:
fusor.angle1() does work better, thanks.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1492,1266 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+    }
+  }
+
+  void TurnDualBladesOn() {
```
This function (and the next one) seem kind of pointless...

NoSloppy replied:
I guess that was from before TurnOn/OffHelper() existed, and then I left them for printouts for testing.
Adjusted.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1492,1266 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+    } else if (mss.x > -14) {
+        thrusting_blade_ = BC_SECOND_BLADE_SET;
+    }
+    thrust_location = EffectLocation(0, thrusting_blade_);
```
Why store this in a variable, why not return it?

NoSloppy replied:
ah yes. 
ok done

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1492,1266 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+
+  bool Event2(enum BUTTON button, EVENT event, uint32_t modifiers) override {
+#ifdef BC_DUAL_BLADES
+    GetThrustBladeLocation();
```
Why calculate this for every event and store it in a variable when it only seems to be used for *one* event?

NoSloppy replied:
fixed.

NoSloppy replied:
no wait, not yet.
It's not one event. there's all blades off, then if one is on.
almost fixed.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1492,1266 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+      case EVENTID(BUTTON_AUX2, EVENT_FIRST_SAVED_CLICK_SHORT, MODE_ON):
+        if (!isSecondBladeOn() && !on_pending_) {
+          TurnBladeOn(BC_SECOND_BLADE_SET);
+        return true;
```
weird indentation

NoSloppy replied:
fixed

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1492,1266 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+// Stab
+      case EVENTID(BUTTON_NONE, EVENT_THRUST, MODE_ON):
+        if (millis() - last_thrust_millis_ < 200 ||
+            millis() - saber_on_time_ < 2000) return false;        
```
move return to next line to make it easier to see

NoSloppy replied:
done

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1492,1266 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+// -------------------- Any btn any blades
+
+// Gesture Ignition Controls
+#ifdef BC_SWING_ON
```
The next chunk of code is hard to read. There are just too many ifdefs.
I have some suggestions below for how to reduce it, but there are other ways, and you don't have to do all of them.

NoSloppy replied:
fixed.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1492,1266 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+#endif
+        if (millis() - saber_off_time_ > 500) {
+          FastOn();
+#ifdef BC_GESTURE_AUTO_BATTLE_MODE
```
How about a helper function: 

void maybe_enable_battle_mode_by_gesture() { 
{
#ifdef BC_GESTURE_AUTO_BATTLE_MODE
  PVLOG_NORMAL << "** Entering Battle Mode\n";
  battle_mode_ = true;
#endif
}

The helper function can then be called without an ifdef.

NoSloppy replied:
Agreed.
Made GestureEnableBattleMode() and NoBladeDisableGestures() helpers and cleaned up this section.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1492,1266 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+          battle_mode_ = true;
+#endif
+#ifdef BC_DUAL_BLADES
+          PVLOG_NORMAL << "** Dual Blades Activated\n";
```
Do you really need this?

NoSloppy replied:
Now that testing is done, no. 
removed.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1492,1266 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
         }
-      }
-      return true;
+        current_menu_angle_ = (fusor.angle2() - (M_PI / 6));
```
This can make current_menu_angle_ less than -M_PI which might not be what you want.
Maybe use the new Angle class for current_menu_angle_ ?

NoSloppy replied:
ah, common/angle.h?
ok so that basically auto wraps to keep it within pi to -pi range.
Would this be the correct syntax then?
`current_menu_angle_ = Angle(fusor.angle2()) - Angle(M_PI / 6);`

profezzorn replied:
That would work, yes. (Although the it also works without the second Angle(), as float will auto-cast to Angle in this case.)

NoSloppy replied:
ok well I've just been messing with this saved twist stuff and the way I have it currently (just updated) seems to work as I intended.
Not resetting the angle after doing EFFECT_USERN, so this specific line goes away.

NoSloppy replied:
Updated saved twist stuff.
Changed variable current_menu_angle to current_twist_angle,
Prevent normal twist after successful TWIST_LEFT/RIGHT gesture for USER effects.
Removed unused defines.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -1234,64 +2775,101 @@ case EVENTID(BUTTON_POWER, EVENT_FOURTH_HELD, MODE_OFF):
-          saber_off_time_ = millis();
-        }
+        if (on_pending_) return;
+        Off(OFF_FAST);
```
This is super weird.
This means that you can turn off the saber from a style with TrDoEffect<>
Is that intentional?
Why is Off() here?

NoSloppy replied:
Hmm. I have no recollection, it's been there a long while.
I mean, being able to turn off the saber from a style with TrDoEffect<> isn't necessarily bad I suppose?
But no, it was not intentional for that purpose.

wow - Looking at it now, I have both FAST_ON and FAST_OFF in SB_Effect. I think I had once assumed they were triggered by saberbase and therefore were a place to set variables etc...
Not right, not good.
So effectively, all of this can go away, right?
```cpp
      // Gesture on, bypass preon
      case EFFECT_FAST_ON:
        if (!SaberBase::IsOn()) FastOn();
#ifdef ENABLE_FASTON
        if (SFX_faston) hybrid_font.PlayPolyphonic(&SFX_faston);
#endif
        scroll_presets_ = false;
        return;
      case EFFECT_FAST_OFF:
        if (on_pending_) return;
        Off(OFF_FAST);
        saber_off_time_ = millis();
        battle_mode_ = false;
        spam_blast_ = false;
        muted_ = false;
        return;
        ```
        
I do not remember the story with faston/fastout really so just removing it all seems like the right thing, especially if it wasn't doing anything anyway.

I tested without this stuff and no difference that I see, so...
removed.

profezzorn reviewed (commented):
Finally made it all the way to the bottom.
I did skim some parts to get there though...
Large code reviews are painful. :(

NoSloppy commented on props/saber_BC_buttons.h:
```diff
@@ -386,36 +1161,248 @@ EFFECT(volmax);     // for maximum volume reached
+    PVLOG_NORMAL << "** Enter Volume Menu\n";
+    initial_volume_ = dynamic_mixer.get_volume();
+    initial_percentage_ = round((initial_volume_ / max_volume_) * 10) * 10;
+   SaberBase::DoEffect(EFFECT_VOLUME_LEVEL, 0);
```
fixed.

NoSloppy commented on props/saber_BC_buttons.h:
```diff
@@ -386,36 +1161,248 @@ EFFECT(volmax);     // for maximum volume reached
+      current_volume_ -= max_volume_ * 0.10;
+      if (current_volume_ < min_volume_) current_volume_ = min_volume_;
+      dynamic_mixer.set_volume(current_volume_);
+      mode::getSL<SPEC>()->SayVolumeDown();
```
announced_ is to prevent it from repeating "Minimum/Maximum Volume" over and over. It says it once and then subsequent adjustments in the same direction say nothing until an opposite adjustment happens.
would it be better to name the variable like range_end_reached_ and/or add a comment explaining?

profezzorn replied:
But you don't need it if the code is changed so that it only says "maximum volume" if the volume 
1. was not at minimum/maximum
2. after incrementing/decrement, it reach minimum/maximum

NoSloppy replied:
Yes ok.
I also moved the actual announce part out of deactivate into its own announce_volume() so it can be called both entering and exiting.
Woking well.

NoSloppy commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1492,1266 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
-          sound_library_.SayDisabled();
+    if (event == EVENT_TWIST) {
+      PVLOG_DEBUG << "**** Detected EVENT_TWIST in Event2, stopping timer and resetting saved twist.\n";
+      saved_twist_ = 0;
```
I don't believe so.
If we use the saved twist in DoSavedTwist(), then it's reset.
But also, if we disregard the saved twist because a normal twist occurred in Event2, it also needs to be reset so that when CheckSavedTwist() is called, it's not lingering... am I wrong?

profezzorn replied:
Right, this is a *normal* twist event.
I was mixing up the events.
Not sure if resetting this here will really fix how twist events interact or not though....

NoSloppy commented:
> Large code reviews are painful. :(

I'll never do one like this again I promise.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -382,40 +1148,252 @@ EFFECT(volup);      // for increse volume
+
+  void QuickMaxVolume() {
+    dynamic_mixer.set_volume(max_volume_);
+    PVLOG_NORMAL << "** Maximum Volume\n";
```
Seems like a lot of these could be PVLOG_VERBOSE

NoSloppy replied:
Could these stay just until finished testing?
Then one clean, small PR could be raising the print log level.
Other verbose messages really clog up the console.

profezzorn replied:
Maybe PVLOG_DEBUG ?

NoSloppy replied:
As mentioned above, for now while still testing this PR, it's helpful to have these show without all the other system messages that print with PROFFIEOS_LOG_LEVEL higher than 300.
My intent is to flip most if not all of these printouts to a higher level PVLOG as one of the last pushes.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -382,40 +1148,252 @@ EFFECT(volup);      // for increse volume
+  }
+
+  bool isTimerExpired() {
+    if (!triggered_) return false;
```
This is super confusing, because isTimerExpired and isTimerRunning sounds very similar, and it seems like they should do basically the same thing. However, isTimerRunning() returns true every time you call it until the timer has expired, but isTimerExpired() only returns true once.

What's worse is that if you call both isTimerRunning and isTImerExpired on repeatedly on the same timer, isTimerExpired might never trigger, because the one time it returned true actually occurred when it was called from isTimerRunning().

NoSloppy replied:
They shouldn't ever get called repeatedly, but after moving volume menu stuff to modes, the only thing left calling isTimerRunning() is DoQuote(), which can just check isTimer NOT Expired instead.
Adjusted and tested. Better now?

NoSloppy replied:
Ok so this needed additional work because it's a little different than the mute and beep timers. Those run a timer to preempt another action, whereas to just prevent overlap for quotes (and force too now) we need to play the sound once to get a duration before starting a timer. So, to bypass checking on the first occurence, I added boolean flag that gets set true once a timer has been used.
I also renamed quote delay timer to overlap_delay_timer since it now has multiple use cases, as well as clarified the other timer names to specify they are used for delaying things.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -425,6 +1403,34 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
   const char* name() override { return "SaberBCButtons"; }
 
+#if defined(DYNAMIC_BLADE_LENGTH) && !defined(MENU_SPEC_TEMPLATE)
+  virtual void EnterBladeLengthMode() {
```
don't make things virtual if they don't need to be (wastes a little bit of memory)

NoSloppy replied:
Not sure why that was ever virtual. Fixed.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1483,1218 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
-      PVLOG_NORMAL << dynamic_mixer.get_volume() << "\n";
-    } else {
-      QuickMaxVolume();
+  void DetectMenuTurn() {
```
It would make a lot more sense to me if this function updated current_twist_angle_

NoSloppy replied:
current_twist_angle_ is updated eventually in DoSavedTwist().
This whole twist thing seems kludgy to me, but it's working.
I have yet to try just implementing your other suggestion 
"add the code that clears the strokes array in the regular twist code instead, that way, we can just add some code that checks if there is a single twist in the strokes array, and if it’s been long enough and no new twist has started. Once that happens, we can emit an event and clear the array. The required max time between the strokes is currently 200ms, so we don’t need to wait any longer than that.

PS: the easiest way to clear the array is probably to add a “UNKNOWN_GESTURE” to the StrokeType enum, and then just set strokes[NELEM(strokes)-1].type to UNKNOWN_GESTURE."

profezzorn replied:
> current_twist_angle_ is updated eventually in DoSavedTwist().

True, but it is much harder to reason about when and how current_twist_angle_ gets updated.
It would make much more sense if this function handled that part.

The kludgy part is having two different twisty systems IMHO. It would be easier to read and understand how it works if there was only one and it did all the things you wanted it to do...

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1483,1218 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+    PVLOG_NORMAL << "Battery Voltage: " << battery_monitor.battery() << "\n";
+    PVLOG_NORMAL << "Battery Percentage: " <<battery_monitor.battery_percent() << "\n";
+    SaberBase::DoEffect(EFFECT_BATTERY_LEVEL, 0);
+    return;
```
reduntant return

NoSloppy replied:
oh not needed. yes, now removed.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1478,1218 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+      if (SetMute(true)) {
+        unmute_on_deactivation_ = true;
+        muted_ = true;
+        TurnOnHelper();
```
SetMute returns false if the sound is already muted, but shouldn't we still call TurnOnHelper() in that case?
(same thing two more times below.)

NoSloppy replied:
I think because SetMute is set back to true on retraction, there's never a chance of the sound already being muted when we get here.
The only time it's already muted would be during a fast_preset change, in which case it just "quickly" remutes.
https://github.com/NoSloppy/ProffieOS-1/blob/0df79d5b72caaa799577914bdced51d40a1bc67f/props/saber_BC_buttons.h#L1658

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -477,738 +1478,1218 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
-    // Scroll Presets timer check - avoid beep/wav overlap
-    if (scroll_presets_timer_.timerCheck() && scroll_presets_) {
+#endif
+    // Allow beep to be heard.
```
This comment is a bit confusing. But I think I get it. (We're delaying playing font.wav so that some beep can be heard first.) Maybe expand the comment a bit?

NoSloppy replied:
Done.

profezzorn commented:
Ok, time to merge this.
It's not perfect, but I can't review everything properly all in one chunk, it's just too much.
In particular, the twist stuff is wonky and should be fixed.
